src/libm/CMakeLists.txt

```diff
@@ -60,6 +60,13 @@ elseif(SLEEF_ARCH_S390X)
     PURECFMA_SCALAR
     DSP_SCALAR
   )
+elseif(SLEEF_ARCH_RISCV64)
+  set(SLEEF_HEADER_LIST
+    RVVM1
+    RVVM2
```

Comment:
Sorry I only noticed that. Doid you not wish to provide *NOFMA versions (and consequently making cinz-variants available to users)? Did you have a reason for that?
AFAICT without this feature users won't be able to have access to a bitwise reproducible implementation across architectures that include both riscv and an architecture without FMA.
You could add it later but I don't think it is much more complicated than what you already have.

Reply:
> AFAICT without this feature users won't be able to have access to a bitwise reproducible implementation across architectures that include both riscv and an architecture without FMA.

I'm not sure I understand what you mean by "include both riscv and an architecture without FMA"? Do you mean if the user is running on a riscv machine with RVV, but that doesn't support FMA? And actually, what is FMA in that context?

Reply:
This option (cinz) is here to provide reproducibility across architectures. Say a user is running on both an old architecture with no support for fused-multiply accumulate (FMA) and a newer FMA-enabled architecture say SVE or RVV, but wants bit reproducible results across those 2 architectures then it won't be possible unless there is a version of the helpers where the FMA is explicitly disabled ({RVV,SVE,...}*NOFMA). That is because RVV will use FMAs in places the old architecture does not (making old arch most likely less accurate, because 1 more rounding per FMA).

Reply:
I recommend implementing it, because many users (including the Unity game-engine) end up using SLEEF because this feature is available. 
This obviously comes at a performance cost, that users are usually happy to pay.

Reply:
Obviously penalty is only on NOFMA and cinz routines, no impact on other routines.

Reply:
There is no RVV without FMA, so on RISC-V only that would be quite artificial. However, in the context of comparing RISC-V to other older architecture, I understand the need for NOFMA. @ericlove you want to tackle that? I can take a look but it's going to take longer :)

Reply:
I just added it with https://github.com/shibatch/sleef/pull/477/commits/395a6f64bdedec5635afebe6cccee230c2290ce3. Let me know if it looks good, it's passing tests but not sure it's done correctly.

Reply:
Thanks @luhenry for adding the NOFMA variants and @blapie for pointing out their absence. I had not added them due to laziness and also not having any need myself for bitwise reproducibility across architectures, but they should of course be there for completeness, and the exercise of adding them seems to have uncovered an important error.

Reply:
WIP at https://github.com/rivosinc/sleef/commits/upstream-riscv-2

Reply:
Fix looking good so far. The use of CONFIG (generally in SLEEF) isn't ideal, but at least it is consistent with other extensions.

Reply:
I'm surprised this wasn't picked up in tests... Could be a combination of things...

Reply:
Fixed with https://github.com/shibatch/sleef/compare/ce3f65a210a04bdb5cec31cd2f384a4e0078987d%5E...e1b59d388395c36a5fa3324fd40973370c47fa78

Reply:
Is your surprise because you think the tests should have reported an error that was above of the specified accuracy tolerance? If so, that error might have been masked by another mistake I made, which was that I did not correctly implement the `vmla*` functions for the NOFMA case, and was emitting FMAs for them anyways.  I see @luhenry has corrected this in his recent commits.

Reply:
Adding support for tester3 is a WIP at https://github.com/rivosinc/sleef/tree/upstream-riscv-2. I got the tests to build and run, but it's failing on a few tests:
```
1/4 Test  #7: tester3rvvm1 .....................***Failed  153.57 sec
sincos u10 7ff4e4fa460faee9b03c42ebb5918f5a
sincos u10 3fe37f4eb805505152f2b14a22a9f94e

sincos u35 5258989dd1f21e6550e4f95d2e7fe46b
sincos u35 95a7b7f48c71febf10ec6eff796dd391

sinf u35 390f777eb73eb2d76914ad4a1d4ff0ba
sinf u35 f8f804eae1d9443103e81fec96293477

cosf u35 03f61afc17485903a0dc002ce0326b5d
cosf u35 f2f3d1c9f090cde9c02439608dc7066e

tanf u35 8ed99e88f25785e580648e649c17fb06
tanf u35 68d42ad1fb412e6b8be3853461e61213


2/4 Test  #4: tester3rvvm1nofma ................***Failed  183.22 sec
sincos u10 5ece9dc5126b287182ea568f53f11382
sincos u10 7c164edcaa45988f6165b653fc76c495

sincos u35 fc3732f1bd033e30a69117f850a1462c
sincos u35 38fe7e261e184ed8dbf432ce6bedc5c4

sinf u35 f45b5ded20a1746474b2f3ddc62c13c1
sinf u35 833d845950b9cbb025629fe4c040f8f6

cosf u35 0fbf8ef44169b5ba1a9e97afd4544fce
cosf u35 74d7f871a6553cd0019087895e2052ad

tanf u35 c19c7d1e49f68be66181934ece099de8
tanf u35 bbb7c092d017e96d2454a38a20687735


3/4 Test #13: tester3rvvm2 .....................***Failed  282.74 sec
sincos u10 7ff4e4fa460faee9b03c42ebb5918f5a
sincos u10 3fe37f4eb805505152f2b14a22a9f94e

sincos u35 5258989dd1f21e6550e4f95d2e7fe46b
sincos u35 95a7b7f48c71febf10ec6eff796dd391

sinf u35 390f777eb73eb2d76914ad4a1d4ff0ba
sinf u35 f8f804eae1d9443103e81fec96293477

cosf u35 03f61afc17485903a0dc002ce0326b5d
cosf u35 f2f3d1c9f090cde9c02439608dc7066e

tanf u35 8ed99e88f25785e580648e649c17fb06
tanf u35 68d42ad1fb412e6b8be3853461e61213


4/4 Test #10: tester3rvvm2nofma ................***Failed  392.52 sec
sincos u10 5ece9dc5126b287182ea568f53f11382
sincos u10 7c164edcaa45988f6165b653fc76c495

sincos u35 fc3732f1bd033e30a69117f850a1462c
sincos u35 38fe7e261e184ed8dbf432ce6bedc5c4

sinf u35 f45b5ded20a1746474b2f3ddc62c13c1
sinf u35 833d845950b9cbb025629fe4c040f8f6

cosf u35 0fbf8ef44169b5ba1a9e97afd4544fce
cosf u35 74d7f871a6553cd0019087895e2052ad

tanf u35 c19c7d1e49f68be66181934ece099de8
tanf u35 bbb7c092d017e96d2454a38a20687735
```

I don't understand enough of SLEEF to figure out where that could come from. @blapie any pointer?

Reply:
Is it failing on all archs? Can you point to the failing GHA, I cannot find it in your fork.

Reply:
I would not blame the riscv implementation of nofma, since most routines are passing, right? Or are they not tested?

Thinking out loud. The common denominator here seems to be trig functions, that 
- all come in DETERMINISTIC and !DETERMINISTIC flavours
- all use rempi(f) that is special-cased for certain vector extensions (however sin,cos,tan and sincosf don't fail...) https://github.com/shibatch/sleef/blame/69cfa7130f3f2bda838633071b1c63aa7d56f940/src/libm/sleefsimddp.c#L298 

What happens if you also special case RISCV(NOFMA) there too? Maybe it has to do with how RISCV flushes to zero or something.

It would be good to see what input fails to confirm that this is the type of issue we are dealing with.

Reply:
That seems to be a riscv64 specific issue: https://github.com/rivosinc/sleef/actions/runs/6893048894. I will try to investigate tomorrow, but any additional pointers would be greatly appreciated.

Reply:
Good news, should be easier to debug. I'll try to have a look too, but I'm not super familiar with risc-v intrinsics.
Looks like the scalar variant passes at least!
I would look at what is specific to trig functions that gets you a different result between scalar and vector (nofma).
Also are the helpers/mappings in tester3.c definitely right, I'd presume so if most tests are passing?

What command line do you use to fetch artifacts, what is the artifact ID?
That didn't work for me
```
gh api -H "Accept: application/vnd.github+json" -H "X-GitHub-Api-Version: 2022-11-28" rivosinc/sleef/actions/artifacts/test-riscv64-llvm-11
```
I would need to see the cmake test log with the details of the failures.

Reply:
I generally just click on the artifacts on the UI, and scp it to my VM if I need.

Reply:
> All tests but 4 do pass, and it was the first thing I checked. Not sure if there are tests for these as well just to make sure I didn't screw it up.

Unfortunately, unit tests isn't really a thing in SLEEF...

> I generally just click on the artifacts on the UI, and scp it to my VM if I need.

Lol, sorry I forgot to scroll.

Reply:
Will have to run with ctest --output-on-failure to get more from log.

Reply:
> Will have to run with ctest --output-on-failure to get more from log.

That's already the case on GHA. And if you want the tests' output from CI, you can download the `test-riscv64-llvm-11` artifact for example.

Reply:
Some values differ in the sincos function. Exhaustive values: https://gist.github.com/luhenry/8d102921ae20ca3a4d9c1270facb1e5c.

I haven't figure out yet why that is, but looking into it.

Reply:
Looking closely into the data, the differences are on -1 on the righest bit. For example,

Reply:
The `vrint_*` were indeed faulty and I've got something passing locally. Testing on CI and pushing on this PR.

Reply:
@luhenry yeah at the time I wrote the original port, there were no intrinsics with an `frm` parameter, so I made this apparently numerically-incorrect hack, but since these have been added now, I'm happy to see your use of them.  Thanks for fixing this!